masp_primitives/src/transaction/components/amount.rs

```diff
@@ -44,71 +57,91 @@ impl memuse::DynamicUsage for Amount {
     }
 }
 
-impl<Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone> Amount<Unit> {
-    /// Returns a zero-valued Amount.
-    pub fn zero() -> Self {
-        Amount(BTreeMap::new())
-    }
-
+impl<Unit, Magnitude> Amount<Unit, Magnitude>
+where
+    Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone,
+    Magnitude: BorshSerialize + BorshDeserialize + PartialEq + Eq + Copy + Default + PartialOrd,
+{
     /// Creates a non-negative Amount from an i64.
     ///
     /// Returns an error if the amount is outside the range `{0..MAX_MONEY}`.
-    pub fn from_nonnegative<Amt: TryInto<i64>>(atype: Unit, amount: Amt) -> Result<Self, ()> {
-        let amount = amount.try_into().map_err(|_| ())?;
-        if amount == 0 {
+    pub fn from_nonnegative(atype: Unit, amount: Magnitude) -> Result<Self, ()> {
+        if amount == Magnitude::default() {
             Ok(Self::zero())
-        } else if 0 <= amount && amount <= MAX_MONEY {
+        } else if Magnitude::default() <= amount {
```

Comment:
Do you think it would be worth having a trait for these magnitudes? Then we could add associated `MIN_MONEY` and `MAX_MONEY` constants.

Reply:
Maybe as a separate PR. For this PR I think it's probably be simpler to not restrict the range of `Amount`s allowed by the MASP crate (beyond what is already imposed by the cryptography).